runtime/lua/vim/lsp.lua

```diff
@@ -1908,6 +1913,239 @@ lsp.commands = setmetatable({}, {
+---                                               is not found then no LSP
+---                                               client will be attached.
+---                         - servers (table): A mapping from filetype to LSP
+---                                   server configuration. The values of this
+---                                   table accept the same keys accepted at the
+---                                   top level of |vim.lsp.config()|.
+---
+function lsp.config(config)
+  return require('vim.lsp.config')(config)
+end
+
+do
+  -- Two-level mapping from filetype -> root dir -> client
+  local client_map = {}
+
+  ---@private
+  local function make_client_config(opts)
+    local config = vim.tbl_deep_extend("keep", opts, lsp.config(), {
+      name = opts.cmd[1],
+    })
+
+    config.on_init = function(client, result)
+      if result.offsetEncoding then
+        client.offset_encoding = result.offsetEncoding
+      end
+
+      if client.config.settings then
+        client.notify("workspace/didChangeConfiguration", {
+          settings = client.config.settings,
+        })
+      end
+
+      if opts.on_init then
+        opts.on_init(client, result)
+      end
+    end
+
+    return config
+  end
+
+  --- Find the first ancestor directory starting from {path} that contains any
+  --- of the file or directory names listed in {markers}.
+  ---
+  --- TODO(gpanders): This function (or something like it) should probably be
+  --- promoted to the stdlib, as it's useful in a variety of situations beyond
+  --- just LSP and is also useful for users/plugins
+  ---
+  ---@param path (string) Start directory
+  ---@param markers (table) A list of file or directory names that are searched
+  ---               for recursively upward.
+  ---@param stop (string) Stop searching when traversal reaches this directory. Defaults to the
+  --             value of $HOME.
+  ---@return (string) The first ancestory directory containing one of the file or
+  ---        directory names listed in {markers}, or nil if no directory was
+  ---        found.
+  ---@private
+  local function find_root(path, markers, stop)
+    validate {
+      path = { path, "s" },
+      markers = { markers, vim.tbl_islist, "a list of root markers" },
+      stop = { stop, "s", true },
+    }
+
+    stop = stop or vim.env.HOME
+
+    local curdir = path
+    while true do
+      for _, v in ipairs(markers) do
+        if vim.fn.globpath(curdir, v) ~= "" then
+          return curdir
+        end
+      end
+
+      local parent = vim.fn.fnamemodify(curdir, ":h")
+      if parent == stop or parent == curdir then
+        return nil
+      end
+      curdir = parent
+    end
+  end
+
+  --- Start (or attach) an LSP client in the current buffer.
+  ---
+  --- Use this function to manually start an LSP client configured with
+  --- |vim.lsp.config()| when "autostart" is not enabled.
+  ---
+  ---@param bufnr (number) buffer number
+  function lsp.buf_start(bufnr)
```

Comment:
I'm wondering how the nvim-jdtls case would work best with this.

I can see two options.

1) Users configure the server via lsp.config, without autostart

```lsp
vim.lsp.config({
    java = {
      ...
      autostart = false
    }
})
```

And setup their own java ft hook that calls `require('jdtls').start_or_attach()`. `jdtls` retrieves the config and then calls `lsp.config` with the updated configuration (not too nice(?)), or it calls `buf_start` with the extended configuration.



2) Users can configure a custom `buf_start`


```lsp
vim.lsp.config({
    java = {
      ...
      buf_start = require('jdtls').start_or_attach
    }
})
```

Then `start_or_attach` would internally extend the configuration and finally call `lsp.buf_start`